File: frontend/src/api/curator.ts
import { apiFetchAuth } from "@/api/client";
import { API_BASE_URL } from "@/lib/env";
import { useAuthStore } from "@/stores/auth";

export interface CuratorSession {
  session_id: string;
  title: string;
  updated_at: string;
}

export async function fetchCuratorSessions(): Promise<CuratorSession[]> {
  const data = await apiFetchAuth<{ sessions: CuratorSession[] }>("/api/v1/curator/sessions");
  return data.sessions;
}

export interface CuratorMessageItem {
  role: "user" | "assistant";
  content: string;
}

export async function fetchSessionMessages(sessionId: string): Promise<CuratorMessageItem[]> {
  const data = await apiFetchAuth<{ messages: CuratorMessageItem[] }>(
    `/api/v1/curator/sessions/${sessionId}/messages`,
  );
  return data.messages;
}

export async function deleteSession(sessionId: string): Promise<void> {
  await apiFetchAuth(`/api/v1/curator/sessions/${sessionId}`, { method: "DELETE" });
}

export interface CuratorSSEEvent {
  event: "token" | "status" | "chart";
  content: string;
}

export async function* streamCurator(
  message: string,
  sessionId?: string,
  imageBase64?: string,
  imageMimeType?: string,
): AsyncGenerator<CuratorSSEEvent> {
  const token = useAuthStore.getState().token;

  const response = await fetch(`${API_BASE_URL}/api/v1/curator/stream`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({
      message,
      session_id: sessionId,
      image_base64: imageBase64,
      image_mime_type: imageMimeType,
    }),
  });

  if (!response.ok || !response.body) {
    throw new Error(`Curator error: ${response.status}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";

    let currentEvent: "token" | "status" = "token";
    for (const line of lines) {
      if (line.startsWith("event: ")) {
        const ev = line.slice(7).trim();
        if (ev === "status" || ev === "token" || ev === "chart") currentEvent = ev as typeof currentEvent;
      } else if (line.startsWith("data: ")) {
        try {
          const data = JSON.parse(line.slice(6)) as { content: string };
          if (data.content) yield { event: currentEvent, content: data.content };
        } catch {
          // ignore malformed lines
        }
      }
    }
  }
}

File: backend/app/schemas/curator.py
"""Curator API Pydantic 스키마."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CuratorChatRequest(BaseModel):
    message: str = Field(default="", description="사용자 질문")
    session_id: str | None = Field(
        default=None, description="대화 세션 ID (없으면 신규)"
    )
    image_base64: str | None = Field(
        default=None, description="이미지 base64 인코딩 데이터"
    )
    image_mime_type: str | None = Field(
        default=None, description="이미지 MIME 타입 (image/jpeg 등)"
    )
    persist: bool = Field(
        default=True,
        description="대화를 DB에 저장할지 여부. false면 히스토리·세션 목록에 남지 않는다 (예: /me 허브 채팅)",
    )


class CuratorSessionItem(BaseModel):
    session_id: str
    title: str
    updated_at: datetime


class CuratorSessionListResponse(BaseModel):
    sessions: list[CuratorSessionItem]


class CuratorMessageItem(BaseModel):
    role: str
    content: str


class CuratorSessionMessagesResponse(BaseModel):
    messages: list[CuratorMessageItem]
